clippy_lints/src/unused_self.rs

```diff
@@ -0,0 +1,102 @@
+    /// of an instance method if it doesn't require `self`.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// struct A;
+    /// impl A {
+    ///     fn method(&self) {}
+    /// }
+    /// ```
+    ///
+    /// Could be written:
+    ///
+    /// ```rust,ignore
+    /// struct A;
+    /// impl A {
+    ///     fn method() {}
+    /// }
+    /// ```
+    pub UNUSED_SELF,
+    complexity,
+    "methods that contain a `self` argument but don't use it"
+}
+
+declare_lint_pass!(UnusedSelf => [UNUSED_SELF]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedSelf {
+    fn check_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &Item) {
+        if item.span.from_expansion() {
+            return;
+        }
+        if let ItemKind::Impl(_, _, _, _, None, _, ref impl_item_refs) = item.kind {
+            for impl_item_ref in impl_item_refs {
+                if_chain! {
+                    if let ImplItemRef {
+                        kind: AssocItemKind::Method { has_self: true },
+                        ..
+                    } = impl_item_ref;
+                    let impl_item = cx.tcx.hir().impl_item(impl_item_ref.id);
+                    if let ImplItemKind::Method(_, body_id) = &impl_item.kind;
+                    then {
+                        let body = cx.tcx.hir().body(*body_id);
+                        let self_param = &body.params[0];
+                        let self_hir_id = self_param.pat.hir_id;
+                        let visitor = &mut UnusedSelfVisitor {
+                            cx,
+                            uses_self: false,
+                            self_hir_id: &self_hir_id,
+                        };
+                        visitor.visit_body(body);
+                        if !visitor.uses_self {
+                            span_help_and_lint(
+                                cx,
+                                UNUSED_SELF,
+                                self_param.span,
+                                "unused `self` argument",
+                                "consider refactoring to a static method or function",
+                            )
+                        }
+                    }
+                }
+            }
+        };
+    }
+}
+
+struct UnusedSelfVisitor<'a, 'tcx> {
+    cx: &'a LateContext<'a, 'tcx>,
+    uses_self: bool,
+    self_hir_id: &'a HirId,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for UnusedSelfVisitor<'a, 'tcx> {
+    fn visit_path(&mut self, path: &'tcx Path, _id: HirId) {
+        if self.uses_self {
+            // This function already uses `self`
+            return;
+        }
+        if let Res::Local(hir_id) = &path.res {
+            self.uses_self = self.self_hir_id == hir_id
+        }
+        walk_path(self, path);
+    }
+
+    fn nested_visit_map<'this>(&'this mut self) -> NestedVisitorMap<'this, 'tcx> {
+        NestedVisitorMap::All(&self.cx.tcx.hir())
```

Comment:
I don't think `All` is required here and that `None` should be enough.

Reply:
If I just change it to `None`, the lint fails to detect the use of `self` in a closure, is there something else I'd need to add?

Reply:
Oh ok, and with [`OnlyBodies`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/intravisit/enum.NestedVisitorMap.html#variant.OnlyBodies)?

Reply:
That seems to work, thanks